src/pybind/mgr/volumes/fs/operations/trash.py

```diff
@@ -61,24 +61,30 @@ def purge(self, trashpath, should_cancel):
         """
         def rmtree(root_path):
             log.debug("rmtree {0}".format(root_path))
+
             try:
                 with self.fs.opendir(root_path) as dir_handle:
                     d = self.fs.readdir(dir_handle)
-                    while d and not should_cancel():
+                    while d and should_purge() and not should_cancel():
```

Comment:
Looks like the way this has been implemented will let the purge threads pickup an entry for synchronisation and then back off as soon as it tries to delete it. I would like to have this for the already processing threads as a way to backoff.

Pausing the purge should be done by using condition variable wait/signal semantics IMO.

Another point about the way `should_purge` lambda is used here - I think it's not required. You could cancel the ongoing deletions using `purge_queue.cance_jobs()`.

Reply:
> You could cancel the ongoing deletions using purge_queue.cance_jobs().

For that, we'll need to pass an instance of `purge_queue` to `operations.trash.Trash.purge()` along with `should_purge()`, so that `purge_queue.cancel_jobs()` can be called when `should_purge()` returns true. Is that correct?

Reply:
Not really - you can invoke that from module.py as `self.vc.purge_queue.cancel_jobs(...)`.

Reply:
> Not really - you can invoke that from module.py as `self.vc.purge_queue.cancel_jobs(...)`.

Okay, it's having no effect. I removed lambda function `should_purge` and called `cancel_all_jobs()` (which is in `purge_queue.py`) as soon as config opt value is changed. 

Trying to figure now why was subvolume purged even though `cancell_all_jobs()` was called.

Reply:
So cancelling all jobs in ineffective and technically buggy.

cancel_all_jobs() empties the vols queue[[1]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L257) and jobs queue[[2]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L266). But then all volumes that are not in the queue will be added back here[[3]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L49) and execution of cancelled jobs will begin here[[4]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L59). This renders `cancel_all_jobs()` ineffective.

Even if we fix this to make sure that cancelled jobs remain cancelled, we'll need to add some more code to re-enable queuing of cancelled jobs when config opt `disable_purge_trash` is turned off. I believe the solution we had on this PR (using lambda function should_purge()) was much simpler.

@vshankar Let me know if I should proceed to fix `cancel_all_jobs()`.

Reply:
> So cancelling all jobs in ineffective and technically buggy.
> 
> cancel_all_jobs() empties the vols queue[[1]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L257) and jobs queue[[2]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L266). But then all volumes that are not in the queue will be added back here[[3]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L49) and execution of cancelled jobs will begin here[[4]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L59). This renders `cancel_all_jobs()` ineffective.
> 
> Even if we fix this to make sure that cancelled jobs remain cancelled, we'll need to add some more code to re-enable queuing of cancelled jobs when config opt `disable_purge_trash` is turned off. I believe the solution we had on this PR (using lambda function should_purge()) was much simpler.
> 
> @vshankar Let me know if I should proceed to fix `cancel_all_jobs()`.

@rishabh-d-dave based on what we discussed, it this still an issue?

Reply:
> So cancelling all jobs in ineffective and technically buggy.
> 
> cancel_all_jobs() empties the vols queue[[1]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L257) and jobs queue[[2]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L266). But then all volumes that are not in the queue will be added back here[[3]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L49) and execution of cancelled jobs will begin here[[4]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L59). This renders `cancel_all_jobs()` ineffective.

[3] only happens when option `periodic_async_work` is enabled. Furthermore, you'd need to wait on a condition (variable) before that.

> 
> Even if we fix this to make sure that cancelled jobs remain cancelled, we'll need to add some more code to re-enable queuing of cancelled jobs when config opt `disable_purge_trash` is turned off.

This would essentially be:

```
self.q = [e['name'] for e in list_volumes(self.vc.mgr)]
```

> I believe the solution we had on this PR (using lambda function should_purge()) was much simpler.
> 
> @vshankar Let me know if I should proceed to fix `cancel_all_jobs()`.

Is there anything to fix? You'd need to do two things:

- Introduce a condition variable (or use an existing one) on which the threads would wait until a certain condition is valid again.
- invoke cancel_all_jobs() which would effectively drain `self.q`. The threads would backoff and block on the condition variable. If subvolumes are deleted from this time onwards, the respective volume gets added to `self.q`, but the threads are still blocked on the condition variable. When the condition becomes valid, refresh `self.q` - the purge operations should start.

Reply:
Yep, I've pushed these changes on this PR some time ago.

Reply:
> Yep, I've pushed these changes on this PR some time ago.

I still see the implementation not using wait-signal semantics - any reason you think its unneeded?

Also, you mentioned about `cancel_all_jobs()` being buggy - is that still the case?